### PR TITLE
Upgrade/rebase onto upstream

### PR DESCRIPTION
**Pull request by atzannes**
# Purpose
Merge in all the upstream changes that happened in the last 4 years because we have been having very regular buildpack breakages due to the code being so old.

Fixes [ALTOS-16888](https://altschool.atlassian.net/browse/ALTOS-16888)

# Approach
First tried to merge the upstream into our master. That rust failed.

Then tried to rebase our master onto upstream, which also lead to lots of extensive conflicts.

Because our changes were few, I reapplied them in separate commits that correspond to the changes we had made on top of that 4-yo upstream/master branch.

The commits that were re-applied are the following:
* f267d3e
* 5e557c4
* f642f1f
* 3518954

The following two commits were *NOT* re-applied because they had been already addressed by the top of upstream/master:
* 28f6d0a
* 3396012

Because our fork was an indirect fork of the official upstream through Mikhail's fork, I modified our repo to be a direct fork of the upstream repo. Github does not allow an account to have multiple forks of the same repo, even if their source (Mikhail vs Heroku) are different. I tried renaming the repo in case the problem was a name conflict, but it wasn't, so I transferred ownership of our fork to my private account (https://github.com/atzannes/heroku-buildpack-python-old), created a fresh AltSchool fork of the Heroku repo, and cherry-picked all the relevant changes from the transferred repo to this one.

# Tests
Pushed to `master-deploy-api` after updating its buildpack to point to this branch and the deploy was successful.

# Pre-Merge TODOs
- [ ] update the buildpack for `master-deploy-api` to point to the master branch

### CODE REVIEW

File: bin/steps/pip-install [hunk 1]
```diff
@@ -25,7 +25,7 @@ if [ ! "$SKIP_PIP_INSTALL" ]; then
     fi
 
 
-    /app/.heroku/python/bin/pip install -r "$BUILD_DIR/requirements.txt" --exists-action=w --src=/app/.heroku/src --disable-pip-version-check --no-cache-dir 2>&1 | tee "$WARNINGS_LOG" | cleanup | indent
+    /app/.heroku/python/bin/pip install -r "$BUILD_DIR/requirements.txt" --exists-action=w --src=/app/.heroku/src --disable-pip-version-check --process-dependency-links --no-cache-dir 2>&1 | tee "$WARNINGS_LOG" | cleanup | indent
```
Comment (jvineet) [inline bin/steps/pip-install:28]: ah, earlier version of pip didn't need the `--process-dependency-links` flag to install private dependencies.

File: bin/release [hunk 2]
```diff
@@ -16,7 +16,7 @@ EOF
 if [[ $MANAGE_FILE ]]; then
 cat <<EOF
 
-addons:
-  - heroku-postgresql
+#addons:
+#  - heroku-postgresql
```
Comment (jvineet) [inline bin/release:20]: adding heroku-postgresql adddon has been commented out, but looks like we were using it earlier.

Comment (atzannes) [reply bin/release:20]: No we weren't :) 
Exhibit A: `bin/release` on the [old master branch](https://github.com/atzannes/heroku-buildpack-python-old/blob/master/bin/release) and on the current [legacy branch](https://github.com/AltSchool/heroku-buildpack-python/blob/legacy/bin/release)

Exhibit B: the [commit](https://github.com/atzannes/heroku-buildpack-python-old/commit/7493d1b06af1703d49a46d29097493ad0398f917) that made this change originally

Comment (jvineet) [reply bin/release:20]: 👍

**Review by jvineet — APPROVED** [review]

**Pull request merged**